File: Day #334/solution.cpp
// 24 Game (fixed order): recursive split of contiguous list, combine with +,-,*,/ (doubles).
// Time: O(1) for fixed 4 numbers. Space: O(1).
#include <bits/stdc++.h>
using namespace std;

vector<double> solve(const vector<double>& a) {
    int n = a.size();
    if (n == 1) return {a[0]};
    vector<double> res;
    for (int i = 1; i < n; ++i) {
        vector<double> L = solve(vector<double>(a.begin(), a.begin() + i));
        vector<double> R = solve(vector<double>(a.begin() + i, a.end()));
        for (double l : L) for (double r : R) {
            res.push_back(l + r);
            res.push_back(l - r);
            res.push_back(l * r);
            if (fabs(r) > 1e-9) res.push_back(l / r);
        }
    }
    return res;
}

bool can24(const vector<double>& a) {
    for (double v : solve(a)) if (fabs(v - 24.0) < 1e-6) return true;
    return false;
}

int main() {
    vector<double> nums = {5, 2, 7, 8};
    cout << (can24(nums) ? "True" : "False") << endl;
    return 0;
}

File: Day #334/solution.rs
// 24 Game (fixed order): recursive split of contiguous list, combine with +,-,*,/ (f64).
// Time: O(1) for fixed 4 numbers. Space: O(1).

fn solve(a: &[f64]) -> Vec<f64> {
    if a.len() == 1 {
        return vec![a[0]];
    }
    let mut res = Vec::new();
    for i in 1..a.len() {
        let l_vals = solve(&a[..i]);
        let r_vals = solve(&a[i..]);
        for &l in &l_vals {
            for &r in &r_vals {
                res.push(l + r);
                res.push(l - r);
                res.push(l * r);
                if r.abs() > 1e-9 {
                    res.push(l / r);
                }
            }
        }
    }
    res
}

fn can24(a: &[f64]) -> bool {
    solve(a).iter().any(|&v| (v - 24.0).abs() < 1e-6)
}

fn main() {
    let nums = [5.0, 2.0, 7.0, 8.0];
    println!("{}", if can24(&nums) { "True" } else { "False" });
}
